feat(Interactions): member objects for uncached guilds

advaith1 opened the pull request:
**Please describe the changes this PR makes and why it should be merged:**
fixes #10010
[Discord thread](https://discord.com/channels/222078108977594368/1179951996841771109)

Currently, when an interaction is received from a guild that isn't cached*, raw API member objects are returned instead of a d.js GuildMember object. This PR creates a new UncachedGuildMember object that is returned instead of raw API member objects.

This new object tries to be compatible with GuildMember, but retains backwards compatibility with APIGuildMember as to not make this a breaking change. Therefore it is still somewhat awkward and not great, but it is much better than just receiving an APIGuildMember. Hopefully in v15 we can remove the backwards compatibility and align it better with GuildMember.

**This has not been fully tested and likely needs changes - please thoroughly review!**

\* This was always possible by adding apps with only the applications.commands scope, but user apps makes this a much more common appearance. This PR is based on the discord.js patch AdvaithBot is running to properly handle user app commands.

**Status and versioning classification:**
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

<!--
Please move lines that apply to you out of the comment:
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- This PR **only** includes non-code changes, like changes to documentation, README, etc.
-->

vladfrangu commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,332 @@
+
+    /**
+     * The guild that this member is part of
+     * @type {Guild}
```
Its a string, not a Guild. You can add in a guild getter

advaith1 replied:
oops will fix

vladfrangu commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,332 @@
+     */
+    this.guildId = guildId;
+
+    if ('user' in data) {
```
Can user ever be missing? 👀

advaith1 replied:
I was going off the member docs where it's optional, but actually as far as we are concerned here it'll always be present. will change that

vladfrangu commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,332 @@
+    /**
+     * Whether the user is deafened in voice channels
+     * @type {?boolean}
+     * @deprecated
```
Why is this deprecated, and what't the alternative?

vladfrangu commented on packages/discord.js/typings/index.d.ts:
```diff
@@ -1642,6 +1642,38 @@ export class GuildMember extends Base {
   public valueOf(): string;
 }
 
+export class UncachedGuildMember extends Base {
```
It doesn't extend just base, you probably need to extend similar to GuildMember/User

advaith1 replied:
the GuildMember class definition also extends base, but that also has an interface so ig that's what's missing? I'm not sure why that has both an interface and a class

vladfrangu replied:
We probably use the interface to make mixins easier, not sure 😓

Qjuh replied:
Yep, it‘s the improved mixin typings using interface merging. The old Mixins were a bad fit for documentation purposes

vladfrangu reviewed (changes requested):
There's a LOT of inconsistencies between trying to make this a non-semver major change (by also having snake_case properties) and a semver-major change (by not exposing all values under the same name). Let's pick one side (and gut says you want this in a sem-minor 👀) and roll with it!

- `nick` isn't exposed as `nick`, only `nickname`
- `data.roles` isn't exposed at all, only as `roleIds`
	- this one I understand why, but I'd at least try to make uncachedMember.roles be our role manager (it _should_ work)

advaith1 commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,332 @@
+      this.avatar === member.avatar &&
+      this.pending === member.pending &&
+      this.communicationDisabledUntilTimestamp === member.communicationDisabledUntilTimestamp &&
+      this.parsedFlags.bitfield === member.parsedFlags.bitfield &&
```
.flags is deprecated so I didn't want to use it; I want to rename parsedFlags to flags in v15

advaith1 commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,332 @@
+ *   .catch(console.error);
+ */
+
+TextBasedChannel.applyToClass(UncachedGuildMember);
```
ah thanks, will look into that

NotSugden commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,332 @@
+    /**
+     * The last timestamp this member started boosting the guild, as an ISO8601 timestamp
+     * @type {?string}
+     * @deprecated Use {@link UncachedGuildMember#premiumSinceTimestamp}
+     * or {@link UncachedGuildMember#premiumSince} instead
```
```suggestion
     * @deprecated Use {@link UncachedGuildMember#premiumSinceTimestamp} or {@link UncachedGuildMember#premiumSince} instead
```

Richy79 replied:
Richardkouakouchangeascommit.12fr

advaith1 replied:
this is what I had originally, but it fails prettier due to the line length, so I split it up.

NotSugden commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,332 @@
+    return this.premiumSinceTimestamp && new Date(this.premiumSinceTimestamp);
+  }
+
+  /**
+   * The member's id
+   * @type {Snowflake}
+   * @readonly
+   */
+  get id() {
+    return this.user.id;
+  }
```
does this need to be a getter. Why can't we just assign it statically as user ids don't change.

tipakA replied:
Because if it isn't a getter you're storing the id twice in each member object

NotSugden commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,332 @@
+
+TextBasedChannel.applyToClass(UncachedGuildMember);
+
+exports.UncachedGuildMember = UncachedGuildMember;
```
is there a reason we don't just export the class directly since its the only thing in this file

advaith1 replied:
idk, GuildMember does the same thing. should I change it here? is there a reason GuildMember does it?

Qjuh replied:
Historical reasons only. GuildMember used to export two things until removed in #7092 
So yeah, you should export as default here

vladfrangu commented:
Marking the PR as both sem minor and major until I test it out and ensure this _is_ a sem-minor PR 👀

Qjuh commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,328 @@
+     * The user that this guild member instance represents
+     * @type {User}
+     */
+    this.user = this.client.users._add(data.user, true);
```
Isn’t this a breaking change? `APIInteractionGuildMember#user` is `APIUser` and thus has snakecase properties. Now it doesn’t anymore.

vladfrangu replied:
Yep, that is a breaking change indeed... @advaith1 can you solve that too?

NotSugden commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,328 @@
+
+  /**
+   * The nickname of this member, or their user display name if they don't have one
+   * @type {?string}
```
would they not always have a _display_ name, or is this something to do with partials?

advaith1 replied:
hm yeah, why is this nullable in GuildMember? also user is nullable in GuildMember but id (which just returns this.user.id) isn't
are both of those just incorrectly marked as nullable

NotSugden replied:
it would seem so

tipakA replied:
Is the user being nullable related to the
> The field `user` won't be included in the member object attached to `MESSAGE_CREATE` and `MESSAGE_UPDATE` gateway events.

note that is under the Guild Member object table [in the DAPI docs](https://discord.com/developers/docs/resources/guild#guild-member-object)?

NotSugden replied:
this sounds like something we need to investigate.... possibly seperately
but that would mean the docs on here could be wrong until investigation
complete

_re posted because email replies dont work properly_

Qjuh replied:
In those two events the `message.author` field contains the `user` and is patched into the `GuildMember` accordingly so won’t be `null` either https://github.com/discordjs/discord.js/blob/de14c92c1158d3e5d7d87d29d2fe9d99eb407df5/packages/discord.js/src/structures/Message.js#L339

NotSugden replied:
so in conclusion: the docs are wrong

NotSugden commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,328 @@
+ * @example
+ * // Send a direct message
+ * UncachedGuildMember.send('Hello!')
+ *   .then(message => console.log(`Sent message: ${message.content} to ${guildMember.displayName}`))
```
```suggestion
 *   .then(message => console.log(`Sent message: ${message.content} to ${uncachedGuildMember.displayName}`))
```
Feels a bit silly but eh

NotSugden commented on packages/discord.js/typings/index.d.ts:
```diff
@@ -1642,6 +1642,46 @@ export class GuildMember extends Base {
+  );
+  public avatar: string | null;
+  public get dmChannel(): DMChannel | null;
+  public get displayName(): string;
```
i note infact from my previous comment that this is marked as always present but in the jsdoc it is not

NotSugden commented:
this sounds like something we need to investigate.... possibly seperately
but that would mean the docs on here could be wrong until investigation
complete

On Wed, 10 Apr 2024, 01:01 tipakA, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In packages/discord.js/src/structures/UncachedGuildMember.js
> <https://github.com/discordjs/discord.js/pull/10195#discussion_r1558486097>
> :
>
> > +  get id() {
> +    return this.user.id;
> +  }
> +
> +  /**
> +   * The DM between the client's user and this member
> +   * @type {?DMChannel}
> +   * @readonly
> +   */
> +  get dmChannel() {
> +    return this.client.users.dmChannel(this.id);
> +  }
> +
> +  /**
> +   * The nickname of this member, or their user display name if they don't have one
> +   * @type {?string}
>
> Is the user being nullable related to the
>
> The field user won't be included in the member object attached to
> MESSAGE_CREATE and MESSAGE_UPDATE gateway events.
>
> note that is under the Guild Member object table in the DAPI docs
> <https://discord.com/developers/docs/resources/guild#guild-member-object>?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/discordjs/discord.js/pull/10195#discussion_r1558486097>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG42MKM7HLSI4URNWHRXYDTY4R6MDAVCNFSM6AAAAABFN3V2DKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSOJQGQ3TGNRWGM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Jiralite commented on packages/discord.js/src/structures/UncachedGuildMember.js:
```diff
@@ -0,0 +1,328 @@
+    return this.nickname ?? this.user.displayName;
+  }
+
+  /**
+   * The nickname of the member
+   * @name UncachedGuildMember#nick
+   * @type {?string}
+   * @deprecated Use {@link UncachedGuildMember#nickname} instead
+   */
+  get nick() {
+    return this.nickname;
+  }
```
Why is this here?

Qjuh replied:
To be non-breaking for https://discord-api-types.dev/api/discord-api-types-v10/interface/APIInteractionGuildMember#nick

vladfrangu reviewed (commented):
Thoughts on an `AnyGuildMember` type which is union of both Uncached and GuildMember and we have functions that assert whether a member instance is the uncached one?

advaith1 commented:
Since this PR targets d.js v15 now, I've done some changes:

- renamed UncachedGuildMember to MinimalGuildMember
- removed backwards compatibility with the API object (since it was incorrect for the user field)
- GuildMember now extends MinimalGuildMember and adds getters/functions that need the guild object
- added member.isInCachedGuild() typeguard that assets it is a GuildMember
